### PR TITLE
Use filerGroup for s3 buckets collection prefix

### PR DESCRIPTION
**Pull request by SmsS4**
# What problem are we solving?
[link to discussion](https://github.com/seaweedfs/seaweedfs/discussions/4463)  
We can't have two buckets with the same name in two different filers with different filer stores because they use the same collection name in the volume server  and of course, we want this change to be backward compatible

# How are we solving the problem?
Use filerGroup as a prefix for buckets collections name (filerGroup\_bucket)  
I added these flags:  
- s3.collectionPrefix in filer  
```shell
weed filer  -filerGroup abc -s3 -s3.collectionPrefix
```
- collectionPrefix and filerGroup in  s3  
```shell
weed s3 -filerGroup abc -collectionPrefix
```
- s3CollectionPrefix in shell
```shell
weed shell -filerGroup abc -s3CollectionPrefix
```
And when s3 gateway uses filerGroup as a prefix of the collection name and sets the collection name query param  
And change the collection of buckets in other files

# How is the PR tested?
with these commands:  
Run two filer, one volume server and, one master  
```shell
go run ./weed master -volumeSizeLimitMB 100 
go run ./weed volume -port 5555 --dir /tmp/abc -max 0
go run ./weed filer -s3 -port 5566 -s3.port 6655 -s3.collectionPrefix -filerGroup first
go run ./weed filer -port 5567 -filerGroup second -defaultStoreDir second-store
go run ./weed s3 -port 6656 -collectionPrefix -filerGroup second -filer localhost:5567
```
`/tmp/a` is a big file  
```shell
echo "s3.bucket.list" | go run ./weed shell -filerGroup first -s3CollectionPrefix
echo "s3.bucket.list" | go run ./weed shell -filerGroup second -s3CollectionPrefix
```
No bucket  
```shell
aws --endpoint http://localhost:6655 s3 mb s3://test
echo "s3.bucket.list" | go run ./weed shell -filerGroup first -s3CollectionPrefix
>   test	size:0	chunk:0
echo "s3.bucket.list" | go run ./weed shell -filerGroup second -s3CollectionPrefix
> 
aws --endpoint http://localhost:6656 s3 mb s3://test
echo "s3.bucket.list" | go run ./weed shell -filerGroup first -s3CollectionPrefix
>   test	size:0	chunk:0
echo "s3.bucket.list" | go run ./weed shell -filerGroup second -s3CollectionPrefix 
>   test	size:0	chunk:0
```
Now we have two bucket with same name and we are going to upload files to theme
```shell
aws --endpoint http://localhost:6655 s3 cp /tmp/a s3://test
aws --endpoint http://localhost:6656 s3 cp /tmp/a s3://test
echo "collection.list" | go run ./weed shell
> collection:""	volumeCount:7	size:186600	fileCount:3	deletedBytes:0	deletion:0
collection:"first_test"	volumeCount:7	size:8025168	fileCount:411	deletedBytes:0	deletion:0
collection:"second_test"	volumeCount:7	size:8025168	fileCount:411	deletedBytes:0	deletion:0
Total 3 collections.
```
Now we are going to enforce quota on first bucket
```shell
echo "s3.bucket.quota -name test -sizeMB 1" | go run ./weed shell -filerGroup first -s3CollectionPrefix 
echo "s3.bucket.quota.enforce -apply" | go run ./weed shell -filerGroup first -s3CollectionPrefix
>   test	size:8025168	quota:1048576	usage:765.34%
    changing bucket test to read only!
echo "s3.bucket.quota.enforce -apply" | go run ./weed shell -filerGroup second -s3CollectionPrefix
>
aws --endpoint http://localhost:6655 s3 cp /tmp/a s3://test
aws --endpoint http://localhost:6656 s3 cp /tmp/a s3://test
> upload failed: ../../../../../tmp/a to s3://test/a 
> upload: ../../../../../tmp/a to s3://test/a                        
```
Only first bucket is readonly

```shell
echo "s3.bucket.list" | go run ./weed shell -filerGroup first -s3CollectionPrefix 
>   test	size:8025168	chunk:411	quota:1048576	usage:765.34%
echo "s3.bucket.list" | go run ./weed shell -filerGroup second -s3CollectionPrefix
>   test	size:16063432	chunk:411
echo "s3.bucket.delete -name test" | go run ./weed shell -filerGroup first -s3CollectionPrefix
>
echo "s3.bucket.list" | go run ./weed shell -filerGroup first -s3CollectionPrefix 
> 
echo "collection.list" | go run ./weed shell
> collection:""	volumeCount:7	size:593272	fileCount:7	deletedBytes:0	deletion:0
collection:"second_test"	volumeCount:7	size:16063432	fileCount:822	deletedBytes:8011551	deletion:411
Total 2 collections.
```

# Checks
- [ ] I have added unit tests if possible.
- [x] I will add related wiki document changes and link to this PR after merging.

### CODE REVIEW

File: weed/command/server.go [hunk 1]
```diff
@@ -147,6 +147,8 @@ func init() {
 	s3Options.auditLogConfig = cmdServer.Flag.String("s3.auditLogConfig", "", "path to the audit log config file")
 	s3Options.allowEmptyFolder = cmdServer.Flag.Bool("s3.allowEmptyFolder", true, "allow empty folders")
 	s3Options.allowDeleteBucketNotEmpty = cmdServer.Flag.Bool("s3.allowDeleteBucketNotEmpty", true, "allow recursive deleting all entries along with bucket")
+	s3Options.filerGroup = cmdServer.Flag.String("filerGroup", "", "filerGroup of filer to use as prefix if collectionPrefix is enable")
```
Comment (chrislusf) [inline weed/command/server.go]: Get `filerGroup` from the response: https://github.com/seaweedfs/seaweedfs/blob/9204ee2d2dfd421753dad9fcc80c2b5ce3ea5326/weed/command/s3.go#L171

File: weed/command/server.go [hunk 2]
```diff
@@ -147,6 +147,8 @@ func init() {
...
+	s3Options.collectionPrefix = cmdServer.Flag.Bool("collectionPrefix", false, "use filerGroup as collection prefix for buckets")
```
Comment (chrislusf) [inline weed/command/server.go]: I prefer to avoid this option and always use filer group as prefix.

File: weed/s3api/filer_util.go [hunk 3]
```diff
@@ -107,6 +107,13 @@ func (s3a *S3ApiServer) updateEntry(parentDirectoryPath string, newEntry *filer_
 	return err
 }
 
+func (s3a *S3ApiServer) getCollectionName(bucket string) string {
+	if s3a.option.CollectionPrefix {
```
Comment (chrislusf) [inline weed/s3api/filer_util.go]: if `filerGroup` is an empty string, use the bucket directly.

File: weed/command/server.go (hunk 2 above)
Comment (SmsS4) [reply weed/command/server.go]: What about when you want to have a filer with filerGroup and s3 enabled?  
for example:  
```
weed filer  -filerGroup abc -s3
```
If we use filerGroup as a prefix in this example our change is not backward compatible  
maybe use `s3.collectionPrefix` but don't use this flag in `weed s3`?

Comment (chrislusf) [reply weed/command/server.go]: S3 instance should get the filer group value from https://github.com/seaweedfs/seaweedfs/blob/9204ee2d2dfd421753dad9fcc80c2b5ce3ea5326/weed/command/s3.go#L171

Comment (chrislusf) [reply weed/command/server.go]: filerGroup is still kind of WIP. It is rarely used I would say. We can do this non-backward compatible change. If customers really want to use this, they can rename existing collections by prefixing the filerGroup to the collection name.

**Comment by SmsS4** [comment]
Looks like there is some network issue with archive.ubuntu.com so End to End / FUSE Mount fails  I event pushed another commit to trigger builds again but again build failed.  
I removed those flags from s3 and shell and get filerGroup from filer response

File: weed/s3api/s3api_object_copy_handlers.go [hunk 4]
```diff
@@ -100,7 +100,7 @@ func (s3a *S3ApiServer) CopyObjectHandler(w http.ResponseWriter, r *http.Request
 	}
 	glog.V(2).Infof("copy from %s to %s", srcUrl, dstUrl)
 	destination := fmt.Sprintf("%s/%s%s", s3a.option.BucketsPath, dstBucket, dstObject)
-	etag, errCode := s3a.putToFiler(r, dstUrl, resp.Body, destination)
+	etag, errCode := s3a.putToFiler(r, dstUrl, resp.Body, destination, dstBucket)
```
Comment (chrislusf) [inline weed/s3api/s3api_object_copy_handlers.go:103]: the change seems wrong here.

**Pull request merged**